### ISSUE TITLE
[RFC] Structured Concurrency

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#6468]
Crystal has an great concurrency model based on Fibers and Channels, which can be used to pass messages around.

Fibers are conceptually pretty simple. If you spawn one, it takes off from the main context and runs concurrently for an indefinite amount of time, depending on what it does and how long it takes to do this. From the perspective of the main control flow, it's essentially *fire and forget*.

Real life problems typically ask for a more sophisticated way of handling concurrent tasks. Sometimes you need to wait for either one, some or all tasks to be finished before continuing the main scope. Error handling in concurrent tasks is also important and the ability to cancel the remaining tasks if others have finished or errored.

Fibers and Channels can be used to implement a model for structured concurrency. Given that this is a pretty common idiom, I'd like to see a generalized implementation in Crystal's stdlib. 

## What we have

`HTTP::Server#listen` uses a custom implementation of a wait group executing a number of tasks simultaneously and waiting for all to finish. Other examples are in the `parallel` macro or `Crystal::Compiler#codegen_many_units`. `parallel` is the only feature of structured concurrency currently available in the stdlib, but it is only suitable for a fixed number of concurrent tasks that are known at compile time.

A more generalized approach would help to make this concept easily re-useable.
It can be implemented based on the existing features that Fiber and Channel provide. The only thing that's missing is a way to deliberatly kill fibers and unwrap their stack (see #3561, and a proposed implementation in #6450).

## Background

I recommend reading the articles referenced below. They both describe a model of structured concurrency which essentially restricts the execution of concurrent tasks to a specific scope and having tools to manage them. This contrasts with the model of `go` (Go) and `spawn` (Crystal) which just fires off a new fiber without caring about it's life cycle. This makes it hard to follow control flow: what happens where and when in which scope.

The main idea of this proposal is to understand that each fiber is limited to the scope it is executed in:

> Every time our control splits into multiple concurrent paths, we want to make sure that they join up again.

This ensures that fibers don't get lost doing whatever stuff they might not even be supposed to do anymore.
I believe this concept can be applied to almost any real-life use case of fibers.
Having a structured flow of control also allows for a proper exception flow. Right now, unhandled exceptions within a fiber are just printed and ignored. When a fiber is scoped to some parent context, an exception can just be propagated there.

## Prototype

I have implemented a simple prototype of a concurrency feature (based on `Fiber.cancel` from #6450). The idea is to have a coordination tool for running fibers, called a `Spindle`. It is used to spawn fibers and ensure to collect them. This particular implementation allows running multiple tasks concurrently and if one of them fails, it cancels all the others. This is of course just an example of behaviour, there are many different ways to react.

The code can be found at: https://gist.github.com/straight-shoota/4437971943bae7000f03fabf3d814a2f

I don't have a concrete proposal how this should be implemented in terms of stdlib API's but the general idea is to provide tools for running tasks concurrently. We could even think about removing unscoped `spawn` (it can be considered harmful after all), but that's not necessarily required and can probably be decided upon later.

## References

* [Notes on Structured Concurrency, or: Go Statement Considered Harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful) by Nathaniel J. Smith
* [Structured Concurrency](http://250bpm.com/blog:71) (by Martin Sústrik)
* [Python Concurrency for Mere Mortals](https://www.youtube.com/watch?v=i-R704I8ySE) (talk) by Nathaniel J. Smith

Some examples of similar libraries:

* Go: [WaitGroup](https://golang.org/pkg/sync/#WaitGroup)
* Rust: [CrossBeam](https://docs.rs/crossbeam/0.3.2/crossbeam/struct.Scope.html)
* C: [libdill](http://libdill.org/structured-concurrency.html)
* Python: [Trio](https://trio.readthedocs.io/en/latest/reference-core.html)

**Comment by asterite** [comment]
Just a note that everything in the std related to concurrency, except `spawn` (that is, `parallel`, `future`, etc.) are just small examples to test whether concurrency work, and in my opinion they should also be deleted and rethought from scratch for a unified way to do concurrent programming in Crystal. I mentioned in the past that [WaitGroup is on the roadmap](https://github.com/crystal-lang/crystal/issues/2366#issuecomment-202682923) (apparently someone edited that wiki and removed that, sigh), it's just that implementing all of these patterns without parallelism already there is almost a waste of time because implementing them with parallelism in mind is totally different, I think (at least if you want it to perform well; for example Go's WaitGroup [doesn't use channels and it's super efficient](https://golang.org/src/sync/waitgroup.go?s=574:898#L10).

**Comment by RX14** [comment]
I'd rather write a "good" `Promise` implementation and then have `Promise.all`/`Promise.first` etc. As far as i'm concerned the abstraction of a fiber with a return value, called a promise or future, is a better and more composable than waitgroups and can still be used as waitgroups pretty easily.

**Comment by straight-shoota** [comment]
Promises are more about communication than structuring. You can do a lot of things in a similar way, but it feels like an inferior solution to me. 

Running two methods concurrently would look like this with a promise:
```cr
promise = future ->do_some_stuff
do_some_other_stuff
promise.get
```

In my example implementation it looks like this:
```cr
concurrent do |spindle|
  spindle.spawn ->do_some_stuff
  spindle.spawn ->do_some_other_stuff
end
```
It's  little bit more verbose, but easier to follow control flow. Especially when it gets more complicated.

(Verbosity could actually be reduced when using `with spindle yield`, but I think I like the expressiveness better.)

One way or the other: a good promise implementation would need a way to cancel a fiber as well, wouldn't it?

**Comment by RX14** [comment]
@straight-shoota you could do it like that or you could use promises like this:
```cr
Promise.all(
  Promise.new ->func,
  Promise.new ->func2
)
```

which is both less verbose and more powerful since you can do

```cr
value1, value2 = Promise.all(
  Promise.new ->func,
  Promise.new ->func2
)
```

very easily to get the function's return values.

Changing it to "first fiber wins" aka a race becomes `s/all/first/`, just changing one function instead of having to refactor your code to use a different concurrency mechanism (consider the difference between a function taking a spindle as a param and returning a promise, thats a big refactor)

**Comment by RX14** [comment]
and JS promises have no way to cancel them, we don't need cancellation for them

**Comment by straight-shoota** [comment]
These `Promise.all` and `Promise.first` are nice, but they only allow a static number of concurrent tasks. This is just like the `parallel` macro and not sufficient for a real-world application where tasks can be dynamically added at runtime (like connection handlers of a network server for example).

**Comment by straight-shoota** [comment]
At least I would want to have a way to cancel a promise/fiber (and do it gracefully) in case its result/effect is no longer required. It shouldn't block resources for unnecessary work.

**Comment by RX14** [comment]
@straight-shoota No, `.all` and `.first` can easily take arrays.

**Comment by straight-shoota** [comment]
How would that work with adding child tasks during runtime? You would have to append them to the array - somehow that could probably be made to work with `.all` and `.first`. But it makes a really ugly API using an array as registry for tasks.

**Comment by RX14** [comment]
Ah I see what you want. We could easily add block-based versions of `Promise.first` and `Promise.all` which do what you want, but I can't help but feel that adding tasks at runtime would be hard to track.

**Comment by RX14** [comment]
I was thinking more of a
```cr
Promise.all do |promises|
  promises << ...
end
```

interface *in addition* to the other promise interfaces to make it do wait groups as well.

I guess it's probably quite a hack though, and better to separate the two concepts.

**Comment by straight-shoota** [comment]
@RX14 That doesn't seem much different from
```cr
concurrent do |spindle|
  spindle.spawn ...
end
```

So we pretty much want the same thing in this regard ;)

**Comment by RX14** [comment]
@straight-shoota yeah I like the concept I was just wondering if we could work it into the "promise" concept for simplicity.

**Comment by stakach** [comment]
I've been writing a Promise library for Crystal Lang and it's almost complete (core implementation complete with specs)
https://github.com/spider-gazelle/promise

It might not be "Structured Concurrency" in the most strict sense however it does simplify coordinating a bunch of async events and it's quite a popular paradigm. Would love to see it in the standard library if that's something you would consider.

I think my implementation is pretty neat in any case:

```crystal
require "promise"

promise = Promise.new(Int32)
result = p.then { |result| result.not_nil! + 100 }.then { |result| "change #{result} type" }
promise.resolve(10)

# Can also grab the value via a future
result.value # => "change 110 type"
```

Any `.then` block can change promise types as it propagates down the chain.
`.catch` blocks can only propagate exceptions and are used to recover values back to the initiating promise value to continue.

Thanks to the Crystal type safety it puts most promise implementations to shame.

**Comment by RX14** [comment]
I don't think a promise library in crystal would look anything at all like a JS promise library. `.then` isn't really required at all. Libraries should expose synchronous APIs, and any uses of `Promise` should really stay internal to the application.

**Comment by stakach** [comment]
Well yeah (exposing synchronous APIs)
Not sure how you escape from using the equivalent of `.then` though it's kind of the je ne sais quoi of promises. Without it all you have is futures.

**Comment by RX14** [comment]
@stakach ah, my terminology was all messed up. I'd like futures in the stdlib and promises perhaps can be a shard built on that.

**Comment by RX14** [comment]
I'm still not sure what the conceptual difference between a promise and future would be in crystal though :thinking:

**Comment by jhass** [comment]
I like to think of promises and futures as either end of a pipe, the promise is where I can put something in and the future is where I can wait and listen for the result.

Now taking this picture, I just described a Channel with a buffer capacity of one that can only be written to once. So perhaps for us Promises would actually be just redundant to channels.

**Comment by stakach** [comment]
Promises are really complimentary to channels. Channels for distributing async work and promises for returning the results.

For instance, I'm working on updating this [influxdb library](https://github.com/aca-labs/influxdb.cr) to the latest version of Crystal lang as the original maintainer doesn't have the time.

The [original version](https://github.com/jeromegn/influxdb.cr/blob/master/src/influxdb/database.cr#L40) just spawns to perform an async request (a channel here would be ideal) - however either way you can never know if the request succeeded and you don't always want to wait around at this point for the response.

A Promise in this case is the perfect solution.
* You can ignore the result like the current implementation
* You can be synchronous if you want to be `Promise#value`
* You can use a call back to handle a success or failure condition

Basically promises provide the ultimate flexibility and are very simple to use

**Comment by RX14** [comment]
Well, regardless of the terminology what I want is a fiber which can return a value, and you can wait for it to complete with a value or error

**Comment by straight-shoota** [comment]
@stakach I don't think a promise is the right tool in your example. The library method `#write` should just be blocking like any  Crystal IO method and directly call `#send_write`. It should be the caller's task to manage concurrency and execute this method in a separate fiber where appropriate. The library simply doesn't need to care about that.

This would be idiomatic Crystal with a simplified API by removing `sync` argument and a dependency on a promise implementation. It's consistent with other IO methods, including others on the same type (`select`, `query`, `drop`) don't seem to have an option to execute asynchronously.

**Comment by stakach** [comment]
@straight-shoota problem is running the write concurrently doesn't work on the current version of Crystal so it has to be via the channel to ensure serial writes.

Basically because the HTTP client response from influx is chunked, crystal yields the current fiber while it waits for IO mid HTTP response cycle.
Then I can start another request while the client is still processing the previous response which leads to weird errors. (not sure if this is an issue with the crystal lang HTTP::Client however if you take that influxdb library, make the minor changes to run on 0.25.1 and run the tests, it blows up) - this is a common issue with fibers and the library should definitely deal with this not the caller.

**Comment by stakach** [comment]
The problem of being called from a different fiber is already solved by using the channel - any fiber can call the code, the channel does the work and the promise returns the result.

Very little complexity vs locks and multiple client instances

**Comment by stakach** [comment]
@RX14 I deliver you "a fiber which can return a value, and you can wait for it to complete with a value or error"
https://github.com/spider-gazelle/promise#simple-concurrency-control

**Comment by stakach** [comment]
@straight-shoota thanks for the advice - I threw a lock around the influxdb HTTP client, you were right, was definitely the way to go.
Also watched the Trio video does seem pretty cool. Look forward to Spindles landing!

**Comment by vinipsmaker** [comment]
> * [Notes on Structured Concurrency, or: Go Statement Considered Harmful](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful) by Nathaniel J. Smith

He overlooks the problem too much and the metaphors between one domain and the other are forced to match his worldview.

> * [Structured Concurrency](http://250bpm.com/blog:71) (by Martin Sústrik)

This article is nice.

Just two notes:

### GC vs systems language

GC languages and systems-languages (i.e. languages where the sole runtime is the execution stack itself and you have to care about pointer and lifetimes) behave differently in such situations.

Thread-start functions (or, fiber-start functions) in GC'ed languages will be closures rather than actual plain functions and, if some value is captured from parent scope, these values won't be destroyed when the parent dies. So we don't usually worry about child threads/fibers overlapping the lifetime of the parent thread/fiber (just like what happens in Go's). If you need deterministic destruction of objects, just be explicit about it. `pthread_cancel` and `pthread_join` will do in C. Extending the use-case to the world beyond C, you'll probably follow Java's approach and change cancellation API to throw exceptions and triggers stack unwinding on the target thread/fiber. If you have C++, there will also be RAII scope guards properly handling cancellation requests (and I bet RAII is already outside of Crystal's scope).

### Don't complicate the state machine

This author _too_ fancies about abstractions. There are two strategies here: graceful vs ungraceful shutdown. Interruption/cancellation API is about graceful shutdown. Don't fancy about timeouts and stuff. Ask the thread/fiber to gracefully shutdown and give it its time.

Want ungraceful shutdown leaking resources and possibly leaving live locks and breaking invariants on data structures? As a rule of thumb this is bad design and you don't want to turn into an idiom culturally seeded on your language. Two approaches here would be:

* Spawn a real process (`os.exit()` won't clean any resources) with a greater level of isolation and kill the process when it fails to satisfy an associated watchdog.
* `kill()` that unschedules the fiber forever as soon as scheduler has a chance. Marked with a BIG BOLD “don't use it if you don't know what you're doing” in the documentation. No fancying about timeout, just a simple API. If you want to give time for the fiber to finish, it's `cancel()` + `timedjoin()` (the only API specifically designed for dirty shutdown here is `kill()` and all others are a fundamental vocabulary that emerged elsewhere).

**Comment by chocolateboy** [comment]
Is this impacted (e.g. made easier, harder, or superseded) in any way by the proposed [MT support](https://github.com/crystal-lang/crystal/pull/8112)? Even if it's orthogonal, as suggested [here](https://github.com/crystal-lang/crystal/pull/7547#issuecomment-473196658), it would be nice to have eyes on the MT proposal with compatibility (or non-incompatibility) with this proposal in mind.

**Comment by straight-shoota** [comment]
@chocolateboy Structured concurrency does not depend on multithreading, but it can easily integrate it. A concurrency context allows to configure very specifically whether tasks can be executed in parallel, how many threads, error handling etc.

**Comment by yxhuvud** [comment]
I suppose only a limited version of this would be possible after 1.0, as a unlimited ability to `spawn` without a lifetime restriction isn't really compatible with the concept.

**Comment by straight-shoota** [comment]
@yxhuvud Not sure I understand.
I'm convinced that when a tool for structured concurrency is available, ideally there should really be no reason to use a random anywhere. Best case scenario would be to avoid that completely. But you should still be able to have both, structured and unstructured fibers.
I don't see a reason why unrestricted spawns wouldn't work with a concept of structured concurrency. Technically, the structure would be build upon the unstructured primitives anyway.
Removing `spawn` as top level method would be possible when there's a different tool for spawning fibers in context. But it's not strictly necessary.

**Comment by asterite** [comment]
The way I see it, `spawn` coupled with `Channel` and `select` are the low-level concurrency primitives in Crystal. In fact, they are the same in Go. I see them similar to Erlang/Elixir `receive` and `!` (send). People don't usually use them, they use abstractions on top of them (like `GenServer`), but from time to time they are useful... especially for building more abstractions.

So we should keep those in Crystal, and build good abstractions on top of them. For instance, actors would be nice.

**Comment by yxhuvud** [comment]
"I don't see a reason why unrestricted spawns wouldn't work with a concept of structured concurrency."

Because it defeats the point. Users should be forced to think about the lifetime of the things they spawn, and it should possible to see using visual inspection of the code what fibers can be running at any given point in the code. Or to quote the referred blogpost referred to above with the (quite telling) title `notes-on-structured-concurrency-or-go-statement-considered-harmful` : 
```
The good news, though, is that these problems can all be solved: Dijkstra showed us how! We need to:

-   Find a replacement for go statements that has similar power, but follows the "black box rule",
-   Build that new construct into our concurrency framework as a primitive, and don't include any form of go statement.
```

The whole point of that whole post is a denouncement of the Go model with unbounded spawns!

And yes, fibers would obviously still be implemented using low level constructs. But probably used as often as `Thread.new` by end-user applications (ie never). The easiest way to create them (`spawn`) should not be the one with the worst long time maintainability properties. Hence my comment about a limited implementation (similar to how C still has `goto`).

**Comment by straight-shoota** [comment]
Okay, I see what you mean. There's no technical reason, but it would make sense to advocate the higher level interface. And frankly, assuming its used in every API, there would be no reason for direct access to the low level interface.

I just feel that can only be achieved as a second step. The first step would be to design and implement a high-level interface for structured concurrency. Then wait for it being adopted. After that, the now (hopefully) unused features can be removed/reduced.

**Comment by raphendyr** [comment]
I would like to add presentation about structured concurrency in context of C++ to this conversation. Memory management doesn't apply to Crystal, but allocated resources could be external, e.g., database connection. Hence, the presentation might help people to see how resource allocation can be done safer with structured concurency.

* [Structured Concurrency: Writing Safer Concurrent Code with Coroutines... - Lewis Baker - CppCon 2019](https://youtu.be/1Wy5sq3s2rg)

In addition, I like how Martin creates connection between object lifetime management and fiber lifetime management (see Martin Sústrik's blog post linked in issue description).

**Comment by ludovicdeluna** [comment]
Interesting. The concept of Executors confuse me. I can understand the wrap of the Job around a Task, but the way the Executors is expressed seems over-complex to me.

I've played with the [Task in C#][1], very close to the concept of _future_ in C++ materialized by the [Async template][2]. For the two, a lightweight process is spread in a thread-pool. We can grab the result later and continue the our work. The way Task is used in C# permit to organize the work (when it start, when we need to wait), perform check on the task and act on failure.

The other way is to have a long running process and use the classical channel / subroutine as in Go or Elixir (with the Mailbox / Job primitives). A bit light for more advanced management, but everything is possible.

A very interesting topic anyway :) Thanks a lot.

[1]: https://docs.microsoft.com/fr-fr/dotnet/api/system.threading.tasks.task?view=net-5.0
[2]: https://thispointer.com/c11-multithreading-part-9-stdasync-tutorial-example

**Comment by cyangle** [comment]
There's also discussion of [go concurrency bugs and structured concurrency](https://trio.discourse.group/t/structured-concurrency-in-golang/174). Some bugs were caused by unbounded go statements in the golang stdlib.

I think structured concurrency should be provided by the language as primitives and the stdlib should always use it.

Kotlin coroutines is a very good implementation of structured concurrency. I think Crystal should have something similar to it.

**Comment by chocolateboy** [comment]
Ruby implementation: [Polyphony](https://digital-fabric.github.io/polyphony/) ([discussion](https://trio.discourse.group/t/structured-concurrency-in-ruby/314))

**Comment by stakach** [comment]
@straight-shoota even Java has Structured Concurrency now. Now there is one reason to use Java over Crystal, how long are we going to allow this?

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/charting-the-route-to-multi-threading-support/7320/1

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/rfc-lazy-objects/8460/5

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/similar-to-asyncio/8744/2

**Comment by crysbot** [comment]
This issue has been mentioned on **Crystal Forum**. There might be relevant details there:

https://forum.crystal-lang.org/t/execution-contexts-and-http-server/8901/2
